velox/functions/sparksql/String.h

```diff
@@ -638,4 +638,155 @@ struct LeftFunction {
+///   by a character in ``replace``. The character in ``replace`` is
+///   corresponding to the character in ``match``. The translation will
+///   happen when any character in ``string`` matching with a character in
+///   ``match``.
+template <typename T>
+struct TranslateFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  folly::F14FastMap<std::string, std::string> unicodeDictionary_;
+  folly::F14FastMap<char, char> asciiDictionary_;
+
+  bool isConstantTranslateDictionary_ = false;
+  bool isConstantTranslateDictionaryBuilt_ = false;
+
+  FOLLY_ALWAYS_INLINE folly::F14FastMap<std::string, std::string>
+  buildUnicodeDictionary(
+      const arg_type<Varchar>& match,
+      const arg_type<Varchar>& replace) {
+    folly::F14FastMap<std::string, std::string> dictionary;
+    int i = 0;
+    int j = 0;
+    while (i < match.size()) {
+      std::string replacePiece;
+      // If match's character size is larger than replace's, the extra
+      // characters in match will be removed from input string.
+      if (j < replace.size()) {
+        int replaceCharLength = utf8proc_char_length(replace.data() + j);
+        replacePiece = std::string(replace.data() + j, replaceCharLength);
+        j += replaceCharLength;
+      }
+      int matchCharLength = utf8proc_char_length(match.data() + i);
+      std::string matchPiece = std::string(match.data() + i, matchCharLength);
+      // Only considers the first occurrence of a character in match.
+      dictionary.emplace(matchPiece, replacePiece);
+      i += matchCharLength;
+    }
+    return dictionary;
+  }
+
+  FOLLY_ALWAYS_INLINE folly::F14FastMap<char, char> buildAsciiDictionary(
+      const arg_type<Varchar>& match,
+      const arg_type<Varchar>& replace) {
+    folly::F14FastMap<char, char> dictionary;
+    int i = 0;
+    for (; i < std::min(match.size(), replace.size()); i++) {
+      char matchChar = *(match.data() + i);
+      char replaceChar = *(replace.data() + i);
+      // Only consider the first occurrence of a character in match.
+      dictionary.emplace(matchChar, replaceChar);
+    }
+    for (; i < match.size(); i++) {
+      char matchChar = *(match.data() + i);
+      dictionary.emplace(matchChar, '\0');
+    }
+    return dictionary;
+  }
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& /*config*/,
+      const arg_type<Varchar>* /*string*/,
+      const arg_type<Varchar>* match,
+      const arg_type<Varchar>* replace) {
+    if (match != nullptr && replace != nullptr) {
+      isConstantTranslateDictionary_ = true;
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& match,
+      const arg_type<Varchar>& replace) {
+    if (!isConstantTranslateDictionary_ ||
+        !isConstantTranslateDictionaryBuilt_) {
```

Comment:
This logic doesn't seem right. It is possible that first batch of data is all-ASCII, but next is not. In this case, on first batch we'll build asciiDictionary and set isConstantTranslateDictionaryBuilt_ to true. On next batch, we'll assume unicodeDictionary_ is built, but it will be empty.

Would be nice to add such a test case as well as reverse (unicode batch followed by ascii batch).

Reply:
@mbasmanova, thanks for your insightful comment. Yes, it indeed has a critical issue in firstly tackling ascii batch, then unicode batch, or reversely. Just fixed by using two independent flags to track whether ascii/unicode dictionary is built. And a simple test was also just added to cover such case. Thanks so much!